### PR TITLE
Add types proposal for component settings override

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
Adds settings in the cluster object to override settings like Replicas and Resources on the individual components

**Release note**:
```release-note
Add support for configuring component settings(Replicas & Resources) via the cluster object
```

### CODE REVIEW

File: api/pkg/crd/kubermatic/v1/cluster.go [hunk 1]
```diff
@@ -87,6 +88,35 @@ type ClusterSpec struct {
...
+}
+
+type ComponentSettings struct {
+	Apiserver         ApiserverSettings         `json:"apiserver"`
+	ControllerManager ControllerManagerSettings `json:"controllerManager"`
+	Scheduler         SchedulerSettings         `json:"scheduler"`
+	Etcd              EtcdSettings              `json:"etcd"`
+}
+
+type ApiserverSettings struct {
```
Comment (alvaroaleman) [inline api/pkg/crd/kubermatic/v1/cluster.go]: Please use one type for all of them as long as they are identical

**Review by alvaroaleman — CHANGES_REQUESTED** [review]

File: api/pkg/crd/kubermatic/v1/cluster.go [hunk 2]
```diff
@@ -87,6 +88,35 @@ type ClusterSpec struct {
 	Pause bool `json:"pause"`
 	// PauseReason is the reason why the cluster is no being managed.
 	PauseReason string `json:"pauseReason,omitempty"`
+
+	// Optional component specific overrides
+	ComponentsOverride ComponentSettings `json:"componentsOverride"`
```
Comment (mrIncompetent) [inline api/pkg/crd/kubermatic/v1/cluster.go:93]: Is `ComponentsOverride` a suitable name?
Or rather `ComponentsConfig`  ?

File: api/pkg/crd/kubermatic/v1/cluster.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/crd/kubermatic/v1/cluster.go]: done with https://github.com/kubermatic/kubermatic/pull/1636/commits/b0943f452bf74838d5a6e77e93e54bb46e211044

File: api/pkg/crd/kubermatic/v1/cluster.go (hunk 2 above)
Comment (alvaroaleman) [reply api/pkg/crd/kubermatic/v1/cluster.go:93]: I like it, as it clearly states that changes here result in non-default settings. Especially if we maybe end up dynamically scaling the ressource settings, the name makes sense as it clearly states this results in a basically hardcoded setting.

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**